### PR TITLE
add nullability annotations to ui/text.dart

### PR DESCRIPTION
**Pull request by yjbanov**
## Related Issues

https://github.com/flutter/flutter/issues/53661

### CODE REVIEW

File: lib/web_ui/lib/src/ui/text.dart [hunk 1]
```diff
@@ -1069,42 +1070,42 @@ class TextRange {
...
-  bool get isCollapsed => start == end;
+  bool/*!*/ get isCollapsed => start == end;
 
   /// Whether the start of this range precedes the end.
-  bool get isNormalized => end >= start;
+  bool/*!*/ get isNormalized => end >= start;
 
   /// The text before this range.
-  String textBefore(String text) {
+  String/*!*/ textBefore(String text) {
```
Comment (zanderso) [inline lib/web_ui/lib/src/ui/text.dart:1090]: Here and below do the parameters need to be annotated?

File: lib/web_ui/lib/src/ui/text.dart [hunk 2]
```diff
@@ -1423,23 +1424,23 @@ abstract class Paragraph {
   /// The [boxHeightStyle] and [boxWidthStyle] parameters must not be null.
   ///
   /// See [BoxHeightStyle] and [BoxWidthStyle] for full descriptions of each option.
-  List<TextBox> getBoxesForRange(int start, int end,
+  List<TextBox/*!*/>/*!*/ getBoxesForRange(int start, int end,
```
Comment (zanderso) [inline lib/web_ui/lib/src/ui/text.dart:1427]: ditto here and below

File: lib/web_ui/lib/src/ui/text.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:1090]: Done.

File: lib/web_ui/lib/src/ui/text.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:1427]: Done.

File: lib/ui/text.dart [hunk 3]
```diff
@@ -360,17 +360,17 @@ class TextDecoration {
   const TextDecoration._(this._mask);
 
   /// Creates a decoration that paints the union of all the given decorations.
-  factory TextDecoration.combine(List<TextDecoration> decorations) {
+  factory TextDecoration.combine(List<TextDecoration/*!*/>/*!*/ decorations) {
```
Comment (zanderso) [inline lib/ui/text.dart:363]: Unrelated question: How does nnbd handle factory constructors that return null?

File: lib/web_ui/lib/src/ui/text.dart [hunk 4]
```diff
@@ -131,7 +131,7 @@ class FontWeight {
   ///
   /// Values for `t` are usually obtained from an [Animation<double>], such as
   /// an [AnimationController].
-  static FontWeight lerp(FontWeight a, FontWeight b, double t) {
+  static FontWeight/*?*/ lerp(FontWeight/*?*/ a, FontWeight/*?*/ b, double/*!*/ t) {
```
Comment (zanderso) [inline lib/web_ui/lib/src/ui/text.dart:134]: `values` doesn't have any null entries:

```suggestion
  static FontWeight/*!*/ lerp(FontWeight/*?*/ a, FontWeight/*?*/ b, double/*!*/ t) {
```

File: lib/web_ui/lib/src/ui/text.dart [hunk 5]
```diff
@@ -902,12 +903,12 @@ class TextBox {
...
   ///
   ///  * [direction], which specifies the text direction.
-  double get end {
+  double/*!*/ get end {
     return (direction == TextDirection.ltr) ? right : left;
   }
 
   @override
-  bool operator ==(dynamic other) {
+  bool/*!*/ operator ==(dynamic other) {
```
Comment (zanderso) [inline lib/web_ui/lib/src/ui/text.dart:911]: Does `dynamic` imply `?`?

File: lib/ui/text.dart (hunk 3 above)
Comment (yjbanov) [reply lib/ui/text.dart:363]: Factories are not allowed to return null any more.

File: lib/web_ui/lib/src/ui/text.dart (hunk 4 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:134]: Right, but there's also:

```dart
    if (a == null && b == null)
      return null;
```

File: lib/web_ui/lib/src/ui/text.dart (hunk 5 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:911]: Yes (but `Object` doesn't)

File: lib/web_ui/lib/src/ui/text.dart (hunk 4 above)
Comment (zanderso) [reply lib/web_ui/lib/src/ui/text.dart:134]: Sorry, I don't follow =) Could you spell it out for me a bit more.

Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:134]: Oh! I was looking at the `lib/ui/text.dart`, where the implementation is as follows:

```dart
  static FontWeight/*?*/ lerp(FontWeight/*?*/ a, FontWeight/*?*/ b, double/*!*/ t) {
    assert(t != null);
    if (a == null && b == null)
      return null;
    return values[lerpDouble(a?.index ?? normal.index, b?.index ?? normal.index, t).round().clamp(0, 8) as int];
  }
```

I should just sync the two so they are consistent. Both should be returning `FontWeight?` though.

Comment (yjbanov) [reply lib/web_ui/lib/src/ui/text.dart:134]: Done.

**Comment by yjbanov** [comment]
The luci-engine failure is due to https://github.com/flutter/flutter/issues/58785. Ignoring it.

**Pull request merged**